codex/erasure/erasure.nim

```diff
@@ -64,236 +70,199 @@ type
     decoderProvider*: DecoderProvider
     store*: BlockStore
 
-  GetNext = proc(): Future[?(bt.Block, int)] {.upraises: [], gcsafe, closure.}
-  PendingBlocksIter* = ref object
-    finished*: bool
-    next*: GetNext
+  EncodingParams = object
+    ecK: int
+    ecM: int
+    rounded: int
+    steps: int
+    blocksCount: int
 
-func indexToPos(self: Erasure, encoded: Manifest, idx, step: int): int {.inline.} =
+func indexToPos(steps, idx, step: int): int {.inline.} =
   ## Convert an index to a position in the encoded
   ##  dataset
   ## `idx`  - the index to convert
   ## `step` - the current step
   ## `pos`  - the position in the encoded dataset
   ##
 
-  (idx - step) div encoded.steps
-
-iterator items*(blocks: PendingBlocksIter): Future[?(bt.Block, int)] =
-  while not blocks.finished:
-    yield blocks.next()
+  (idx - step) div steps
 
 proc getPendingBlocks(
   self: Erasure,
   manifest: Manifest,
-  start, stop, steps: int): ?!PendingBlocksIter =
+  indicies: seq[int]): AsyncIter[(?!bt.Block, int)] =
   ## Get pending blocks iterator
   ##
 
   var
-    # calculate block indexes to retrieve
-    blockIdx = toSeq(countup(start, stop, steps))
-    # request all blocks from the store
-    pendingBlocks = blockIdx.mapIt(
-      self.store.getBlock(manifest[it]) # Get the data blocks (first K)
+    # request blocks from the store
+    pendingBlocks = indicies.map( (i: int) =>
+      self.store.getBlock(BlockAddress.init(manifest.treeCid, i)).map((r: ?!bt.Block) => (r, i)) # Get the data blocks (first K)
     )
-    indices = pendingBlocks # needed so we can track the block indices
-    iter = PendingBlocksIter(finished: false)
-
-  trace "Requesting blocks", pendingBlocks = pendingBlocks.len
-  proc next(): Future[?(bt.Block, int)] {.async.} =
-    if iter.finished:
-      trace "No more blocks"
-      return none (bt.Block, int)
-
-    if pendingBlocks.len == 0:
-      iter.finished = true
-      trace "No more blocks - finished"
-      return none (bt.Block, int)
-
-    let
-      done = await one(pendingBlocks)
-      idx = indices.find(done)
-
-    logScope:
-      idx = idx
-      blockIdx = blockIdx[idx]
-      manifest = manifest[blockIdx[idx]]
 
-    pendingBlocks.del(pendingBlocks.find(done))
-    without blk =? (await done), error:
-      trace "Failed retrieving block", err = $error.msg
-      return none (bt.Block, int)
+  proc isFinished(): bool = pendingBlocks.len == 0
 
-    trace "Retrieved block"
-    some (blk, blockIdx[idx])
-
-  iter.next = next
-  success iter
+  proc genNext(): Future[(?!bt.Block, int)] {.async.} = 
+    let completedFut = await one(pendingBlocks)
+    pendingBlocks.del(pendingBlocks.find(completedFut))
```

Comment:
find can return `-1`, which will crash `del`

Reply:
This issue was already prior to my changes. But I will fix it though - I think the best handling would be to return a failed future from this method - this will stop the iterator and then it will propagate the error via `Result.failure` back to the caller of decode/encode.